File: internal/cli/config_test.go
package cli

import (
	"errors"
	"io"
	"os"
	"path/filepath"
	"testing"

	"alpine/internal/apperr"
)

func noSpriteLookup() (string, error) { return "", nil }

func TestSlugify(t *testing.T) {
	got := slugify("Feat/My.Change Test")
	if got != "feat-my-change-test" {
		t.Fatalf("slugify mismatch: %q", got)
	}
}

func TestResolveTokenPrefersEnv(t *testing.T) {
	dir := t.TempDir()
	authPath := filepath.Join(dir, "auth.json")
	if err := os.WriteFile(authPath, []byte(`{"token":"from-file"}`), 0o600); err != nil {
		t.Fatal(err)
	}
	token, warn, err := resolveTokenWithSpriteLookup([]string{"SPRITES_TOKEN=from-env"}, authPath, noSpriteLookup)
	if err != nil {
		t.Fatal(err)
	}
	if token != "from-env" {
		t.Fatalf("expected env token, got %q", token)
	}
	if !warn {
		t.Fatal("expected warning for token mismatch")
	}
}

func TestResolveTokenFallsBackToAuthJSON(t *testing.T) {
	dir := t.TempDir()
	authPath := filepath.Join(dir, "auth.json")
	if err := os.WriteFile(authPath, []byte(`{"token":"from-file"}`), 0o600); err != nil {
		t.Fatal(err)
	}

	token, warn, err := resolveTokenWithSpriteLookup(nil, authPath, noSpriteLookup)
	if err != nil {
		t.Fatal(err)
	}
	if token != "from-file" {
		t.Fatalf("expected file token, got %q", token)
	}
	if warn {
		t.Fatal("did not expect mismatch warning")
	}
}

func TestResolveTokenMissingEverywhere(t *testing.T) {
	dir := t.TempDir()
	authPath := filepath.Join(dir, "auth.json")
	if err := os.WriteFile(authPath, []byte(`{"other":"value"}`), 0o600); err != nil {
		t.Fatal(err)
	}

	_, _, err := resolveTokenWithSpriteLookup(nil, authPath, noSpriteLookup)
	if err == nil {
		t.Fatal("expected missing token error")
	}
}

func TestResolveTokenFallsBackToSpritesLogin(t *testing.T) {
	dir := t.TempDir()
	authPath := filepath.Join(dir, "auth.json")
	if err := os.WriteFile(authPath, []byte(`{"other":"value"}`), 0o600); err != nil {
		t.Fatal(err)
	}

	token, warn, err := resolveTokenWithSpriteLookup(nil, authPath, func() (string, error) {
		return "from-sprites-login", nil
	})
	if err != nil {
		t.Fatal(err)
	}
	if token != "from-sprites-login" {
		t.Fatalf("expected sprites login token, got %q", token)
	}
	if warn {
		t.Fatal("did not expect mismatch warning")
	}
}

func TestTokenFromAuthJSONNestedAuthObject(t *testing.T) {
	dir := t.TempDir()
	authPath := filepath.Join(dir, "auth.json")
	if err := os.WriteFile(authPath, []byte(`{"auth":{"access_token":"nested-token"}}`), 0o600); err != nil {
		t.Fatal(err)
	}

	token, err := tokenFromAuthJSON(authPath)
	if err != nil {
		t.Fatal(err)
	}
	if token != "nested-token" {
		t.Fatalf("expected nested token, got %q", token)
	}
}

func TestParsePlainModeSkipsRepoChecks(t *testing.T) {
	home := t.TempDir()
	t.Setenv("HOME", home)
	authPath := filepath.Join(home, ".local", "share", "opencode", "auth.json")
	configDir := filepath.Join(home, ".config", "opencode")
	if err := os.MkdirAll(filepath.Dir(authPath), 0o755); err != nil {
		t.Fatal(err)
	}
	if err := os.WriteFile(authPath, []byte(`{"token":"from-file"}`), 0o600); err != nil {
		t.Fatal(err)
	}
	if err := os.MkdirAll(configDir, 0o755); err != nil {
		t.Fatal(err)
	}

	cfg, err := Parse([]string{"--plain"}, []string{"SPRITES_TOKEN=from-env"}, io.Discard)
	if err != nil {
		t.Fatalf("expected parse success, got error: %v", err)
	}
	if !cfg.Plain {
		t.Fatal("expected plain mode")
	}
	if cfg.NamePrefix != "plain" {
		t.Fatalf("expected plain name prefix, got %q", cfg.NamePrefix)
	}
	if cfg.LocalEnvFile != "" {
		t.Fatalf("expected no local env file in plain mode, got %q", cfg.LocalEnvFile)
	}
	if cfg.RepoRoot != "" || cfg.RepoURL != "" || cfg.RepoName != "" {
		t.Fatalf("expected repo fields to be empty in plain mode: root=%q url=%q name=%q", cfg.RepoRoot, cfg.RepoURL, cfg.RepoName)
	}
}

func TestParsePlainRejectsBranch(t *testing.T) {
	home := t.TempDir()
	t.Setenv("HOME", home)
	authPath := filepath.Join(home, ".local", "share", "opencode", "auth.json")
	configDir := filepath.Join(home, ".config", "opencode")
	if err := os.MkdirAll(filepath.Dir(authPath), 0o755); err != nil {
		t.Fatal(err)
	}
	if err := os.WriteFile(authPath, []byte(`{"token":"from-file"}`), 0o600); err != nil {
		t.Fatal(err)
	}
	if err := os.MkdirAll(configDir, 0o755); err != nil {
		t.Fatal(err)
	}

	_, err := Parse([]string{"--plain", "--branch", "feat/test"}, []string{"SPRITES_TOKEN=from-env"}, io.Discard)
	if !errors.Is(err, apperr.ErrUsage) {
		t.Fatalf("expected usage error, got: %v", err)
	}
}

File: test/cli/config.test.ts
import { describe, it, beforeEach } from "node:test";
import * as assert from "node:assert/strict";
import * as fs from "node:fs";
import * as path from "node:path";
import * as os from "node:os";
import { parse, slugify, resolveTokenWithSpriteLookup, tokenFromAuthJSON } from "../../src/cli/config.js";
import { AppError } from "../../src/apperr/errors.js";

const noSpriteLookup = (): string | null => null;

function makeTmpDir(): string {
  return fs.mkdtempSync(path.join(os.tmpdir(), "config-test-"));
}

describe("slugify", () => {
  it("converts mixed case with separators", () => {
    const got = slugify("Feat/My.Change Test");
    assert.equal(got, "feat-my-change-test");
  });
});

describe("resolveToken", () => {
  it("prefers env token with warning", () => {
    const dir = makeTmpDir();
    const authPath = path.join(dir, "auth.json");
    fs.writeFileSync(authPath, JSON.stringify({ token: "from-file" }));

    const [token, warn] = resolveTokenWithSpriteLookup(["SPRITES_TOKEN=from-env"], authPath, noSpriteLookup);
    assert.equal(token, "from-env");
    assert.equal(warn, true);
  });

  it("falls back to auth.json", () => {
    const dir = makeTmpDir();
    const authPath = path.join(dir, "auth.json");
    fs.writeFileSync(authPath, JSON.stringify({ token: "from-file" }));

    const [token, warn] = resolveTokenWithSpriteLookup([], authPath, noSpriteLookup);
    assert.equal(token, "from-file");
    assert.equal(warn, false);
  });

  it("errors when missing everywhere", () => {
    const dir = makeTmpDir();
    const authPath = path.join(dir, "auth.json");
    fs.writeFileSync(authPath, JSON.stringify({ other: "value" }));

    assert.throws(
      () => resolveTokenWithSpriteLookup([], authPath, noSpriteLookup),
      (err: unknown) => err instanceof AppError && err.kind === "auth",
    );
  });

  it("falls back to sprites login", () => {
    const dir = makeTmpDir();
    const authPath = path.join(dir, "auth.json");
    fs.writeFileSync(authPath, JSON.stringify({ other: "value" }));

    const [token, warn] = resolveTokenWithSpriteLookup([], authPath, () => "from-sprites-login");
    assert.equal(token, "from-sprites-login");
    assert.equal(warn, false);
  });
});

describe("tokenFromAuthJSON", () => {
  it("reads nested auth object", () => {
    const dir = makeTmpDir();
    const authPath = path.join(dir, "auth.json");
    fs.writeFileSync(authPath, JSON.stringify({ auth: { access_token: "nested-token" } }));

    const token = tokenFromAuthJSON(authPath);
    assert.equal(token, "nested-token");
  });
});

describe("parse", () => {
  let origHome: string | undefined;
  let home: string;

  beforeEach(() => {
    origHome = process.env["HOME"];
    home = makeTmpDir();
    process.env["HOME"] = home;

    const authDir = path.join(home, ".local", "share", "opencode");
    fs.mkdirSync(authDir, { recursive: true });
    fs.writeFileSync(path.join(authDir, "auth.json"), JSON.stringify({ token: "from-file" }));

    const configDir = path.join(home, ".config", "opencode");
    fs.mkdirSync(configDir, { recursive: true });
  });

  // Restore HOME after each test
  // Note: node:test beforeEach doesn't have afterEach paired, so we use a manual approach
  function restoreHome() {
    if (origHome !== undefined) {
      process.env["HOME"] = origHome;
    }
  }

  it("plain mode skips repo checks", () => {
    const devNull = { write: () => true } as unknown as NodeJS.WritableStream;
    const cfg = parse(["--plain"], ["SPRITES_TOKEN=from-env"], devNull);
    assert.equal(cfg.plain, true);
    assert.equal(cfg.namePrefix, "plain");
    assert.equal(cfg.localEnvFile, "");
    assert.equal(cfg.repoRoot, "");
    assert.equal(cfg.repoURL, "");
    assert.equal(cfg.repoName, "");
    restoreHome();
  });

  it("plain rejects branch", () => {
    const devNull = { write: () => true } as unknown as NodeJS.WritableStream;
    assert.throws(
      () => parse(["--plain", "--branch", "feat/test"], ["SPRITES_TOKEN=from-env"], devNull),
      (err: unknown) => err instanceof AppError && err.kind === "usage",
    );
    restoreHome();
  });
});
